everestrs/everestrs/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+    #[error("invalid argument to command call: '{0}'")]
+    InvalidArgument(&'static str),
+}
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+#[cxx::bridge]
+mod ffi {
+    struct CommandMeta {
+        implementation_id: String,
+        name: String,
+        obj: *const c_char,
+    }
+
+    struct JsonBlob {
+        data: Vec<u8>,
+    }
+
+    unsafe extern "C++" {
+        include!("everestrs_sys/everestrs_sys.hpp");
+
+        type Module;
+        fn create_module(module_id: &str, prefix: &str, conf: &str) -> UniquePtr<Module>;
+
+        /// Connects to the message broker and launches the main everest thread to push work
+        /// forward. Returns the module manifest.
+        fn initialize(self: Pin<&mut Module>) -> JsonBlob;
+
+        /// Returns the interface definition.
+        fn get_interface(self: &Module, interface_name: &str) -> JsonBlob;
+
+        /// Registers the callback of the `GenericModule` to be called and calls
+        /// `Everest::Module::signal_ready`.
+        unsafe fn signal_ready(
+            self: &Module,
+            obj: *const c_char,
+            on_ready: unsafe fn(obj: *const c_char) -> (),
+        );
+
+        /// Informs the runtime that we implement the command described in `meta` and registers the
+        /// `handle_command` method from the `GenericModule` as the handler.
+        unsafe fn provide_command(
+            self: &Module,
+            meta: CommandMeta,
+            on_ready: unsafe fn(&CommandMeta, JsonBlob) -> JsonBlob,
+        );
+    }
+}
+
+impl ffi::JsonBlob {
+    fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+
+    fn deserialize<T: DeserializeOwned>(self) -> T {
+        // TODO(hrapp): Error handling
+        serde_json::from_slice(self.as_bytes()).unwrap()
+    }
+
+    fn from_vec(data: Vec<u8>) -> Self {
+        Self { data }
+    }
+}
+
+#[derive(FromArgs, Debug)]
+/// An everest Node.
+struct Args {
+    /// prefix of installation.
+    #[argh(option)]
+    #[allow(unused)]
+    pub prefix: PathBuf,
+
+    /// configuration yml that we are running.
+    #[argh(option)]
+    #[allow(unused)]
+    pub conf: PathBuf,
+
+    /// module name for us.
+    #[argh(option)]
+    pub module: String,
+}
+
+/// Implements the handling of commands & variables, but has no specific information about the
+/// details of the current module, i.e. it deals with JSON blobs and strings as command names. Code
+/// generation is used to build the concrete, strongly typed abstractions that are then used by
+/// final implementors.
+pub trait GenericModule: Sync {
+    /// Handler for the command `name` on `implementation_id` with the given `parameters`. The return value
+    /// will be returned as the result of the call.
+    fn handle_command(
+        &mut self,
+        implementation_id: &str,
+        name: &str,
+        parameters: HashMap<String, serde_json::Value>,
+    ) -> Result<serde_json::Value>;
+
+    fn on_ready(&mut self) {}
```

Comment:
Not sure but I don't expect that the rust compiler can check if someone is holding multiple mutable refs to this struct - since these are called from within the c++ code. I would suggest to change the signature to `&self` such that if users are changing data in the callback they are forced to wrap it in a mutex